### DISCUSSION TITLE
Documentation

### DISCUSSION THREAD

**Discussion by Gaasduif** [leolabs/ableton-js#58] [General]
Hello,

I'd like to have some more documentation about using this module.
Like first, how do I for instance set the tempo etc.... Can't find any info about setting stuff in Ableton, only reading by adding listeners.
I thought I knew some js but reading/debugging this code is overwhelming, probably too little coding knowledge.

**Comment by leolabs** [comment]
Hey @CrossmodGlennVS,

I've just expanded the example code in the readme to include setting a property. I hope this helps, let me know if you have any further questions :)

**Reply by Gaasduif** [reply]
Thanks. 

Is there a way to contact you?
I've got some questions but I don't to overload this project with discussions. 
Some are probably noob questions and such.

**Reply by leolabs** [reply]
Don't worry about asking your questions here – they might be useful to other users as well :)

**Reply by Gaasduif** [reply]
Allright.

At the moment I'm trying to get 'available_input_routing_channels'. 
It's not in the 'track.d.ts'-file, but I do get a response (string ''<Track.RoutingChannelVector object at 0x117289d50>)when trying to 'get' it.
What does this mean? How do I get the object?

**Reply by leolabs** [reply]
If the property is not defined in the `track.d.ts` file, it's likely that it hasn't been implemented yet by this library. Simple values like strings and numbers could be fetched without support by this library, but more complex properties like an array of channels needs proper support to work unfortunately.

I can check if I can implement support for the property you're trying to access. Are there any others you need that currently aren't supported by Ableton.js?

**Reply by Gaasduif** [reply]
Yes, Another one that gave me the samelike response was the volume. I was trying to access it through 'mixer_device' because it's not part of 'track'. So I tried to get 'mixer_device' from a track.

Another question is about listeners:
I have listeners for track-properties (name, color, volume, input/output routing, ...) on all tracks. Does this take lots of processing power? Or do u need to be careful of not having too many listeners in my project?
And if a track gets deleted, do I need to 'unlisten' that listener to clear up memory space?

**Reply by leolabs** [reply]
> Another one that gave me the same response was the volume. I was trying to access it through 'mixer_device' because it's not part of 'track'. So I tried to get 'mixer_device' from a track.

Thanks, I'll also see if I can implement support for that!

> I have listeners for track properties (name, color, volume, input/output routing, ...) on all tracks. Does this take lots of processing power? Or do you need to be careful not to have too many listeners in my project?

You should be alright with adding lots of listeners. Some of my projects use ~10k listeners across cues, tracks, and clips, and it doesn't seem to impact performance.

> And if a track gets deleted, do I need to 'unlisten' that listener to clear up memory space?

You don't have to remove the listeners of removed entities, Live will take care of it automatically.

**Comment by Gaasduif** [comment]
Thank you for replying. I'm here with another question:

I have an array containing all the tracks (got by getting song/'tracks').
Next I'm trying to get the selected track getting property song-view/'selected_track'.
But the nsid of the track that's returned has is different that nsid of that track in my array.

So has the id changed? Or does the id represent something else?

**Reply by leolabs** [reply]
Unfortunately, while one NSID always points to the same object, one object can have multiple NSIDs. This is a limitation from Ableton's API which creates a new reference to a requested object each time it is requested.

Your best bet would be to check if the track's color and name are the same, or if the track's `is_part_of_selection` property is true.

**Reply by Gaasduif** [reply]
So there's also no track# or scene# to identify clipslots ? In Max For Live this could be retrieved when u had an id.

**Reply by leolabs** [reply]
Unfortunately no, I couldn't find any index property on Tracks and ClipSlots. You can take a look at all available properties here: https://nsuspray.github.io/Live_API_Doc/11.0.0.xml

**Reply by leolabs** [reply]
@CrossmodGlennVS I think I have found a way to get a stable ID for objects from Live. Each object has a `_live_ptr` property that doesn't change over the lifespan of a session, and I've used that to generate a stable `id` field on the transmitted objects. I've pushed a new version of ableton-js, could you check if this works for you?

**Reply by Gaasduif** [reply]
I'm getting error : "The installed version of your AbletonJS plugin (2.5.3) is lower than the JS library (2.7.0). Please update your AbletonJS plugin to the latest version: https://git.io/JvaOu"

I've used "npm i ableton-js" to install

**Reply by leolabs** [reply]
You might have to install the latest version of the control script in Live: https://github.com/leolabs/ableton-js#prerequisites

**Reply by Gaasduif** [reply]
That's right!

After reloading the midi scripts I get following exception:
"The command " + cls + "." + name + "(" + arg + ") timed out after " + timeout + " ms.". 
At line 270 of "index.js", but I don't see it.

**Reply by leolabs** [reply]
Hmm, that's strange. It sounds a bit like the control script isn't working properly. Have you made sure it is enabled in Live's MIDI settings? If it's enabled, you can also check Live's log file to see if it contains any errors from the script.

**Reply by Gaasduif** [reply]
Allright. It works! Selecting session clips triggers the listeners for selected track and/or scene.

**Reply by leolabs** [reply]
I think this is just how Live behaves, since selecting a clip also selects the track and scene in the UI.

**Comment by Gaasduif** [comment]
What's the right way to 'getNotes' from a selected clip? I can't seem to find to way to do it. It's not a property, you have to 'sendCommand' it, I believe.

**Reply by leolabs** [reply]
Each clip object has a `getNotes()` method that you can call to get all notes in a certain time and pitch range that you can specify. Here's the method definition: https://github.com/leolabs/ableton-js/blob/master/src/ns/clip.ts#L256-L270

**Reply by Gaasduif** [reply]
I got the clip through: 
this.ab.getProp("clip_slot", (id of clipslot), "clip")

But this gives an object:
{
  is_audio_clip: false,
  name: '',
  color: 16773172,
  is_midi_clip: true,
  start_time: 1624,
  muted: false,
  id: 'live_4388710628',
  end_time: 4
}
it doesn't have any methods.
I think I'm using your libraries the wrong way.

How do I get the class object?

**Reply by leolabs** [reply]
Here's an example of how to get the notes of the clip in the first clip slot of the first track:

```ts
const tracks = await ableton.song.get("tracks");
const clipSlots = await tracks[0].get("clip_slots");
const clip = await clipSlots[0].get("clip");
const notes = await clip?.getNotes(
  0 /* start time */,
  0 /*lowest pitch*/,
  100 /* end time */,
  120 /* highest pitch */,
);
```

**Reply by Gaasduif** [reply]
Thanks! This works. It also makes start my project all over again haha

**Reply by leolabs** [reply]
That's great to hear! Hope it doesn't take too long to change your project :)

**Reply by Gaasduif** [reply]
The problem is that I stubble upon new js stuff that I didnt know before, so I have to learn it and realise that are better ways to reach the result I want.

First I created my own Track-, Scene-, Clipslot-, Clip-classes which hold all properties retrieved by getters and listeners. But these objects are already there in your code. So now I'm thinking of working with your objects and add properties and methods to these objects. So I've just about prototypes. This opens up new insights. 

I might still not be on the right path. Work in progress.

**Comment by Gaasduif** [comment]
Is it possible to add listener to notes from a clip? 
In Max4Live it's possible, but it's not in the ObservableProperties list.
It's not giving an error either.

**Reply by leolabs** [reply]
Oh, maybe I missed the property. There's a listener function for notes in Live's API, so it should be possible. Let me see if I can add it to the library.

**Reply by leolabs** [reply]
I've just released an update that should allow you to add a listener for all notes. Remember to update the python script in Live as well for this to work :)

**Reply by Gaasduif** [reply]
Thanks. It workes. It returns an array instead of an object like with 'getNotes'. Not a problem though.

**Reply by leolabs** [reply]
Hmm, that's strange. It should return an array of note objects when the notes change. I'll check this out soon, in the meantime, you can use the `tupleToNote` function from `ableton-js/util/note` to convert a note array into a note object.
